File: programmers/150365.cpp
#include <string>
#include <vector>
#include <cmath>
using namespace std;

int dx[] = { 1, 0, 0, -1 };
int dy[] = { 0, -1, 1, 0 };
char dir[] = { 'd', 'l', 'r', 'u' };

string answer;
bool chk[55][55][2525];

bool f(int n, int m, int x, int y, int r, int c, int k, int cnt, string path) {
    
    if (cnt == k) {
        if (x == r && y == c) {
            answer = path;
            return true;
        }
        return false;
    }
    
    if (chk[x][y][cnt]) return false;
    chk[x][y][cnt] = true;
    
    int dist = abs(x - r) + abs(y - c);
    if (dist > k || (k - cnt - dist) % 2) return false;

    for (int i = 0; i < 4; i++) {
        int nx = x + dx[i];
        int ny = y + dy[i];

        if (nx < 1 || n < nx || ny < 1 || m < ny) continue;

        if (f(n, m, nx, ny, r, c, k, cnt + 1, path + dir[i])) return true;
    }
    return false;
}

string solution(int n, int m, int x, int y, int r, int c, int k) {
    
    answer = "impossible";
    f(n, m, x, y, r, c, k, 0, "");
    
    return answer;
}

File: programmers/150365.java
import java.util.*;

class Solution {
    private static final int[] DX = { 1, 0, 0, -1 };
    private static final int[] DY = { 0, -1, 1, 0 };
    private static final char[] DIR = { 'd', 'l', 'r', 'u' };
    
    private boolean[][][] visited;
    
    public String solution(int n, int m, int x, int y, int r, int c, int k) {
        visited = new boolean[n + 1][m + 1][k + 1];
        
        Queue<State> q = new ArrayDeque<>();
        q.offer(new State(x, y, 0, ""));
        
        while (!q.isEmpty()) {
            State cur = q.poll();
            
            // 방문 검사
            if (visited[cur.x][cur.y][cur.count]) continue;
            visited[cur.x][cur.y][cur.count] = true;
            
            // k번 이동한 경우
            if (cur.count == k) {
                if (cur.x == r && cur.y == c) {
                    return cur.path;
                }
                continue;
            }
            
            // 4방향 BFS
            for (int d = 0; d < 4; d++) {
                int nx = cur.x + DX[d];
                int ny = cur.y + DY[d];
                
                if (nx < 1 || n < nx || ny < 1 || m < ny) continue;
                
                String npath = cur.path + DIR[d];
                q.offer(new State(nx, ny, cur.count + 1, npath));
            }
        }
        
        return "impossible";
    }
    
    private static class State {
        final int x;
        final int y;
        final int count;
        String path;
        
        State(int x, int y, int count, String path) {
            this.x = x;
            this.y = y;
            this.count = count;
            this.path = path;
        }
    }
}